evm/src/fuzz/invariant/mod.rs

```diff
@@ -0,0 +1,270 @@
+//! Fuzzing support abstracted over the [`Evm`](crate::Evm) used
+use crate::{fuzz::*, CALLER};
+
+mod executor;
+mod filters;
+
+use ethers::{
+    abi::{Abi, Function},
+    prelude::ArtifactId,
+    types::{Address, Bytes, U256},
+};
+use parking_lot::RwLock;
+pub use proptest::test_runner::Config as FuzzConfig;
+use proptest::{
+    option::weighted,
+    strategy::SBoxedStrategy,
+    test_runner::{TestError, TestRunner},
+};
+
+use std::{borrow::BorrowMut, cell::RefMut, collections::BTreeMap, sync::Arc};
+
+use crate::executor::{Executor, RawCallResult};
+
+use proptest::strategy::{Strategy, ValueTree};
+
+pub type TargetedContracts = BTreeMap<Address, (String, Abi, Vec<Function>)>;
+pub type FuzzRunIdentifiedContracts = Arc<RwLock<TargetedContracts>>;
+
+/// (Sender, (TargetContract, Calldata))
+pub type BasicTxDetails = (Address, (Address, Bytes));
+
+/// Metadata on how to run invariant tests
+#[derive(Debug, Clone, Copy, Default)]
+pub struct InvariantTestOptions {
+    /// The number of calls executed to attempt to break invariants in one run.
+    pub depth: u32,
+    /// Fails the invariant fuzzing if a revert occurs
+    pub fail_on_revert: bool,
+    /// Allows randomly overriding an external call when running invariant tests
+    pub call_override: bool,
+}
+
+/// Wrapper around any [`Executor`] implementor which provides fuzzing support using [`proptest`](https://docs.rs/proptest/1.0.0/proptest/).
+///
+/// After instantiation, calling `fuzz` will proceed to hammer the deployed smart contracts with
+/// inputs, until it finds a counterexample sequence. The provided [`TestRunner`] contains all the
+/// configuration which can be overridden via [environment variables](https://docs.rs/proptest/1.0.0/proptest/test_runner/struct.Config.html)
+pub struct InvariantExecutor<'a> {
+    pub executor: &'a mut Executor,
+    /// Proptest runner.
+    runner: TestRunner,
+    /// Contracts deployed with `setUp()`
+    setup_contracts: &'a BTreeMap<Address, (String, Abi)>,
+    /// Contracts that are part of the project but have not been deployed yet. We need the bytecode
+    /// to identify them from the stateset changes.
+    project_contracts: &'a BTreeMap<ArtifactId, (Abi, Vec<u8>)>,
+}
+
+/// Given the executor state, asserts that no invariant has been broken. Otherwise, it fills the
+/// external `failed_invariant` map and returns a generic error.
+pub fn assert_invariants<'a>(
+    test_contract_abi: &Abi,
+    executor: &'a RefCell<&mut &mut Executor>,
```

Comment:
this looks a bid wild lol

it should be possible to only accept `&mut Exec` here and deref/borrow should be handled at the caller site